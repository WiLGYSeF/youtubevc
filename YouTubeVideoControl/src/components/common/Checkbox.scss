label.cpt-checkbox {
  cursor: pointer;
  margin-right: 4px;

  input[type="checkbox"] {
    cursor: pointer;
    position: relative;
    top: 2px;
  }
}
