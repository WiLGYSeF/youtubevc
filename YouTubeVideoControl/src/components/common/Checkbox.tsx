import React, { ChangeEvent } from 'react';

import '../../css/style.min.css';

interface CheckboxProps {
  label: string;
  labelRight?: boolean;

  defaultChecked?: boolean;

  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
}

function Checkbox(props: CheckboxProps) {
  const labelRight = props.labelRight ?? true;
  const defaultChecked = props.defaultChecked ?? false;

  const inputIdInternal = `checkbox-${Math.random().toString(36).substring(2)}`;

  const eLabel = (<span>{props.label}</span>);

  return (
    <label htmlFor={inputIdInternal} className="cpt-checkbox" data-label-right={labelRight}>
      {!labelRight && eLabel}
      <input type="checkbox"
        id={inputIdInternal}
        onChange={props.onChange}
        defaultChecked={defaultChecked}
      />
      {labelRight && eLabel}
    </label>
  );
}

export default Checkbox;
