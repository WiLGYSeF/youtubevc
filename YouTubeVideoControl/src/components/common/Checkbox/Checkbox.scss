label.cpt-checkbox {
  cursor: pointer;
  margin-right: 4px;
  user-select: none;

  input[type="checkbox"] {
    cursor: pointer;
    position: relative;
    top: 2px;
  }
}
