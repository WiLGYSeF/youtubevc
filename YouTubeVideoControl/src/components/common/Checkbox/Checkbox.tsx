import React, { ChangeEvent, useState } from 'react';

import useStatePropBacked from 'utils/useStatePropBacked';

import './Checkbox.scss';

interface CheckboxProps {
  label: string;
  labelLeft?: boolean;

  checked?: boolean;
  onChange(checked: boolean): void;
}

// exported for mocking
export function getIdInternal(): string {
  return `checkbox-${Math.random().toString(36).substring(2)}`;
}

function Checkbox(props: CheckboxProps) {
  const labelLeft = props.labelLeft ?? false;
  const [checked, setChecked] = useStatePropBacked(props.checked ?? false);
  const [inputIdInternal] = useState(getIdInternal());

  const eLabel = (<span>{props.label}</span>);

  return (
    <label htmlFor={inputIdInternal} className="cpt-checkbox" data-label-left={labelLeft}>
      {labelLeft && eLabel}
      <input
        type="checkbox"
        id={inputIdInternal}
        checked={checked}
        onChange={(e: ChangeEvent<HTMLInputElement>) => {
          const isChecked = e.target.checked;
          props.onChange(isChecked);
          setChecked(isChecked);
        }}
      />
      {!labelLeft && eLabel}
    </label>
  );
}

export default Checkbox;
