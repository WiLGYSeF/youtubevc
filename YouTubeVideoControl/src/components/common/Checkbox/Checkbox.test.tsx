import React from 'react';
import { render } from '@testing-library/react';

import flatten from 'utils/flattenElements';
import Checkbox, { getInputs } from './Checkbox';

function getLabel(container: HTMLElement): HTMLElement {
  return container.querySelector('[data-testid="label"]') as HTMLElement;
}

describe('Checkbox', () => {
  let getIdInternalMock: jest.SpyInstance;

  beforeEach(() => {
    getIdInternalMock = jest.spyOn(Checkbox.prototype, 'getIdInternal')
      .mockImplementation(() => 'checkbox-test');
  });

  afterEach(() => {
    getIdInternalMock.mockRestore();
  });

  it('renders label', () => {
    const { container } = render(<Checkbox
      label="test label"
      onChange={() => {}}
    />);

    const label = getLabel(container);
    const { checkbox } = getInputs(container);

    const flattened = flatten(container);

    expect(flattened.indexOf(checkbox)).toBeLessThan(flattened.indexOf(label));
  });

  it('renders label on left', () => {
    const { container } = render(<Checkbox
      label="test label"
      labelLeft
      onChange={() => {}}
    />);

    const label = getLabel(container);
    const { checkbox } = getInputs(container);

    const flattened = flatten(container);

    expect(flattened.indexOf(label)).toBeLessThan(flattened.indexOf(checkbox));
  });

  it('renders checked', () => {
    const { container } = render(<Checkbox
      label="test label"
      onChange={() => {}}
      defaultChecked
    />);

    const { checkbox } = getInputs(container);

    expect(checkbox.checked).toBeTruthy();
  });

  it('renders unchecked', () => {
    const { container } = render(<Checkbox
      label="test label"
      onChange={() => {}}
      defaultChecked={false}
    />);

    const { checkbox } = getInputs(container);

    expect(checkbox.checked).toBeFalsy();
  });

  it('sends value on change', () => {
    let result = false;
    const testFn = jest.fn();

    const { container } = render(<Checkbox
      label="test label"
      onChange={(checked: boolean) => {
        result = checked;
        testFn(checked);
      }}
    />);

    const { checkbox } = getInputs(container);

    expect(checkbox.checked).toBeFalsy();
    expect(result).toBeFalsy();

    checkbox.click();

    expect(checkbox.checked).toBeTruthy();
    expect(result).toBeTruthy();

    checkbox.click();

    expect(checkbox.checked).toBeFalsy();
    expect(result).toBeFalsy();

    expect(testFn.mock.calls.map((x) => x[0])).toEqual([true, false]);
  });

  it('updates value on props change', () => {
    let checked = false;
    const { container, rerender } = render(<Checkbox
      label="test label"
      defaultChecked={checked}
      onChange={() => {}}
    />);

    const { checkbox } = getInputs(container);

    expect(checkbox.checked).toBeFalsy();

    checked = true;
    rerender(<Checkbox
      label="test label"
      defaultChecked={checked}
      onChange={() => {}}
    />);

    expect(checkbox.checked).toBeTruthy();
  });
});

describe('Checkbox getIdInternal', () => {
  it('getIdInternal', () => {
    expect(Checkbox.prototype.getIdInternal()).toMatch(/^checkbox-[A-Za-z0-9]+$/);
  });
});
