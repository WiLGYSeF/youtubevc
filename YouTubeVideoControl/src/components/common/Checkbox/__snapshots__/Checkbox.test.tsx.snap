// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders checked 1`] = `
<label
  className="cpt-checkbox"
  data-label-left={false}
  htmlFor="checkbox-test"
>
  <input
    checked={true}
    id="checkbox-test"
    onChange={[Function]}
    type="checkbox"
  />
  <span>
    test label
  </span>
</label>
`;

exports[`Checkbox renders label 1`] = `
<label
  className="cpt-checkbox"
  data-label-left={false}
  htmlFor="checkbox-test"
>
  <input
    checked={false}
    id="checkbox-test"
    onChange={[Function]}
    type="checkbox"
  />
  <span>
    test label
  </span>
</label>
`;

exports[`Checkbox renders label on left 1`] = `
<label
  className="cpt-checkbox"
  data-label-left={true}
  htmlFor="checkbox-test"
>
  <span>
    test label
  </span>
  <input
    checked={false}
    id="checkbox-test"
    onChange={[Function]}
    type="checkbox"
  />
</label>
`;

exports[`Checkbox renders unchecked 1`] = `
<label
  className="cpt-checkbox"
  data-label-left={false}
  htmlFor="checkbox-test"
>
  <input
    checked={false}
    id="checkbox-test"
    onChange={[Function]}
    type="checkbox"
  />
  <span>
    test label
  </span>
</label>
`;
